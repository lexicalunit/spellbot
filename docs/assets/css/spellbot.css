.hero {
  padding: 1rem;
  text-align: center;
}

.add-bot {
  background-color: rgb(90, 62, 253);
  border-radius: 3px;
  border: none;
  color: rgb(255, 255, 255);
  cursor: pointer;
  font-size: large;
  font-weight: 200;
  line-height: 1;
  min-height: 32px;
  padding: 10px 30px;
  transition: all 0.3s ease-in-out;
  user-select: none;
}

.add-bot:hover {
  filter: brightness(150%);
}

.screenshot {
  border-radius: 1rem;
  background: linear-gradient(45deg, rgb(246, 173, 75), rgb(90, 62, 253));
  padding: 0.5rem;
  box-shadow: 10px 10px rgba(113, 128, 150, 0.2);
  display: block;
  margin-left: auto;
  margin-right: auto;
}

.avatar-img {
  -webkit-animation: spin 60s linear infinite;
  -moz-animation: spin 60s linear infinite;
  animation: spin 60s linear infinite;
}
@-moz-keyframes spin {
  100% {
    -moz-transform: rotate(360deg);
  }
}
@-webkit-keyframes spin {
  100% {
    -webkit-transform: rotate(360deg);
  }
}
@keyframes spin {
  100% {
    -webkit-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

p.note {
  background: linear-gradient(
    45deg,
    rgba(246, 173, 75, 0.2),
    rgba(90, 62, 253, 0.2)
  );
  border-radius: 1rem;
  margin: 2rem 1rem;
  padding: 0.5rem 1rem 0.5rem 3rem;
}

p.note:before {
  content: "ðŸ’¡";
  float: left;
  margin-left: -2rem;
}

.avatar-img-border {
  box-shadow: none !important;
  transition: all 0.3s ease-in-out;
}

.avatar-img-border:hover {
  filter: brightness(150%);
}

a {
  transition: all 0.3s ease-in-out;
}

footer {
  border-top: 0 !important;
}

a:hover {
  text-decoration: none !important;
}

*:not(.fas):not(.fab) {
  font-family: "Verdana", sans-serif !important;
}

.navbar-brand {
  display: none !important;
}

.nav-item img {
  box-shadow: none !important;
  transition: all 0.3s ease-in-out;
}

.nav-item img:hover {
  filter: brightness(150%);
}

code {
  font-size: 1em !important;
  background-color: transparent !important;
}

@keyframes fadeIn {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}

#main-navbar .navbar-nav .nav-item:last-child img {
  animation: fadeIn ease-in 2s;
}
