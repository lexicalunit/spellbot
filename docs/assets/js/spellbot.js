var SpellBotJS = {
  init : function() {
    $('#main-navbar .navbar-nav').append('<li class="nav-item"><a class="nav-link" href="https://top.gg/bot/725510263251402832"><img src="https://top.gg/api/widget/status/725510263251402832.svg?noavatar=true" alt="SpellBot" /></a></li>');
  },
};

document.addEventListener('DOMContentLoaded', SpellBotJS.init);
