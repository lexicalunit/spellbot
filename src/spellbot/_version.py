from __future__ import annotations

from typing import Optional

import dunamai as _dunamai


def version_from_git() -> Optional[str]:
    try:
        choice = _dunamai.Version.from_any_vcs
        version = _dunamai.get_version("spellbot", first_choice=choice).serialize()
        return version if version is not None and version != "0.0.0" else None
    except Exception:
        return None


def version_from_package() -> Optional[str]:
    try:
        choice = _dunamai.Version.from_any_vcs
        version = _dunamai.get_version("spellbot", third_choice=choice).serialize()
        return version if version is not None and version != "0.0.0" else None
    except Exception:
        return None


def version_from_toml() -> Optional[str]:
    from os.path import realpath
    from pathlib import Path

    import toml

    try:
        pkg_root = Path(realpath(__file__)).parent
        src_root = Path(pkg_root).parent
        repo_root = src_root.parent
        pyproject = toml.load(repo_root / "pyproject.toml")
    except Exception:
        return None

    tool = pyproject.get("tool", {})
    poetry = tool.get("poetry", {})
    version = poetry.get("version", None)
    return version if version is not None and version != "0.0.0" else None


# Below we attempt to detect the version from git first, and if
# that fails we fallback to getting it from the project, and if
# that fails we'll just try and read it directly from pyproject.toml.
# If we can successfully get it from git, we can get a development
# version name like 5.0.2.post2.dev0+1d15510 for example.

__version__ = version_from_git()
if __version__ is None:
    __version__ = version_from_package()
if __version__ is None:
    __version__ = version_from_toml()
if __version__ is None:
    __version__ = "unknown"
