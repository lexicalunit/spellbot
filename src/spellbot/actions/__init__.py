from __future__ import annotations

from .admin_action import AdminAction
from .block_action import BlockAction
from .leave_action import LeaveAction
from .lfg_action import LookingForGameAction
from .score_action import ScoreAction
from .tasks_action import TasksAction
from .verify_action import VerifyAction
from .watch_action import WatchAction

__all__ = [
    "AdminAction",
    "BlockAction",
    "LeaveAction",
    "LookingForGameAction",
    "ScoreAction",
    "TasksAction",
    "VerifyAction",
    "WatchAction",
]
