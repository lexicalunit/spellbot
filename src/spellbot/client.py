from __future__ import annotations

import asyncio
import logging
from collections.abc import AsyncGenerator
from contextlib import asynccontextmanager
from typing import Optional
from uuid import uuid4

import discord
from ddtrace import tracer
from discord.ext.commands import AutoShardedBot, CommandError, CommandNotFound, Context
from expiringdict import ExpiringDict

from .database import db_session_manager, initialize_connection
from .metrics import setup_ignored_errors, setup_metrics
from .services import ChannelsService, GuildsService, VerifiesService
from .settings import Settings
from .spelltable import generate_link
from .utils import user_can_moderate

logger = logging.getLogger(__name__)


class SpellBot(AutoShardedBot):
    def __init__(
        self,
        mock_games: bool = False,
        create_connection: bool = True,
    ):
        self.settings = Settings()
        intents = discord.Intents().default()
        intents.members = True  # pylint: disable=E0237
        intents.message_content = True  # pylint: disable=E0237
        logger.info(f"{intents.value=}")
        super().__init__(
            command_prefix="!",
            help_command=None,
            intents=intents,
            application_id=self.settings.BOT_APPLICATION_ID,
        )
        self.mock_games = mock_games
        self.create_connection = create_connection
        self.channel_locks = ExpiringDict(max_len=100, max_age_seconds=3600)  # 1 hr
        self.event

    async def on_ready(self):
        logger.info("client ready")

    async def setup_hook(self) -> None:
        # Note: In tests we create the connection using fixtures.
        if self.create_connection:  # pragma: no cover
            logger.info("initializing database connection...")
            await initialize_connection("spellbot-bot")

        # register persistent views
        from .views import PendingGameView, SetupView, StartedGameView

        self.add_view(PendingGameView(self))
        self.add_view(SetupView(self))
        self.add_view(StartedGameView(self))

        # load all cog extensions and application commands
        from .utils import load_extensions

        await load_extensions(self)

    @asynccontextmanager
    async def channel_lock(self, channel_xid: int) -> AsyncGenerator[None, None]:
        if not self.channel_locks.get(channel_xid):
            self.channel_locks[channel_xid] = asyncio.Lock()
        async with self.channel_locks[channel_xid]:  # type: ignore
            yield

    @tracer.wrap()
    async def create_spelltable_link(self) -> Optional[str]:
        if self.mock_games:
            return f"http://exmaple.com/game/{uuid4()}"
        return await generate_link()

    @tracer.wrap(name="interaction", resource="on_message")
    async def on_message(self, message: discord.Message):  # pylint: disable=arguments-differ
        span = tracer.current_span()
        if span:  # noqa
            setup_ignored_errors(span)

        # handle DMs normally
        if not message.guild or not hasattr(message.guild, "id"):
            return await super().on_message(message)
        if span:  # noqa
            span.set_tag("guild_id", message.guild.id)  # type: ignore

        # ignore everything except messages in text channels
        if not hasattr(message.channel, "type") or message.channel.type != discord.ChannelType.text:
            return
        if span:  # noqa
            span.set_tag("channel_id", message.channel.id)  # type: ignore

        # ignore hidden/ephemeral messages
        if message.flags.value & 64:
            return

        # to verify users we need their user id
        if not hasattr(message.author, "id"):
            return

        message_author_xid = message.author.id  # type: ignore
        if span:  # noqa
            span.set_tag("author_id", message_author_xid)

        # don't try to verify the bot itself
        if self.user and message_author_xid == self.user.id:  # pragma: no cover
            return

        async with db_session_manager():
            await self.handle_verification(message)

    async def on_command_error(  # pylint: disable=arguments-differ
        self,
        context: Context[SpellBot],
        exception: CommandError,
    ) -> None:
        if isinstance(exception, CommandNotFound):
            return
        return await super().on_command_error(context, exception)

    @tracer.wrap()
    async def handle_verification(self, message: discord.Message):
        message_author_xid = message.author.id  # type: ignore
        verified: Optional[bool] = None
        guilds = GuildsService()
        await guilds.upsert(message.guild)
        channels = ChannelsService()
        channel_data = await channels.upsert(message.channel)
        if channel_data["auto_verify"]:
            verified = True
        verify = VerifiesService()
        assert message.guild
        guild: discord.Guild = message.guild  # type: ignore
        await verify.upsert(guild.id, message_author_xid, verified)
        if not user_can_moderate(message.author, guild, message.channel):
            user_is_verified = await verify.is_verified()
            if user_is_verified and channel_data["unverified_only"]:
                await message.delete()
            if not user_is_verified and channel_data["verified_only"]:
                await message.delete()


def build_bot(mock_games: bool = False, create_connection: bool = True) -> SpellBot:
    bot = SpellBot(mock_games=mock_games, create_connection=create_connection)
    setup_metrics()
    return bot
