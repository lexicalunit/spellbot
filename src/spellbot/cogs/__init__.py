import logging
from importlib import import_module
from inspect import isclass
from pathlib import Path
from pkgutil import iter_modules

from discord.ext import commands
from discord.ext.commands import AutoShardedBot

from .about_cog import AboutCog
from .admin_cog import AdminCog
from .block_cog import BlockCog
from .config_cog import ConfigCog
from .events_cog import EventsCog
from .leave_cog import LeaveGameCog
from .lfg_cog import LookingForGameCog
from .owner_cog import OwnerCog
from .score_cog import ScoreCog
from .sync_cog import SyncCog
from .tasks_cog import TasksCog
from .verify_cog import VerifyCog
from .watch_cog import WatchCog

logger = logging.getLogger(__name__)

# Only exported cogs will be loaded into the bot at runtime.
__all__ = [
    "AboutCog",
    "AdminCog",
    "BlockCog",
    "ConfigCog",
    "EventsCog",
    "LeaveGameCog",
    "LookingForGameCog",
    "OwnerCog",
    "ScoreCog",
    "SyncCog",
    "TasksCog",
    "VerifyCog",
    "WatchCog",
]


async def load_all_cogs(bot: AutoShardedBot) -> AutoShardedBot:
    # iterate through the modules in the current package
    package_dir = Path(__file__).resolve().parent
    for info in iter_modules([str(package_dir)]):

        # import the module and iterate through its attributes
        module = import_module(f"{__name__}.{info.name}")
        for attribute_name in dir(module):  # pragma: no cover
            attribute = getattr(module, attribute_name)

            # Only load cogs in this module if they're exported
            if (
                isclass(attribute)
                and issubclass(attribute, commands.Cog)
                and attribute.__name__ in __all__
            ):
                if module.__name__ in bot.extensions:
                    logger.info("reloading extension %s...", module.__name__)
                    await bot.reload_extension(module.__name__)
                else:
                    logger.info("loading extension %s...", module.__name__)
                    await bot.load_extension(module.__name__)
                break
    return bot
