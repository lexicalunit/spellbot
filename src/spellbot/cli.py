# pylint: disable=too-many-arguments
from __future__ import annotations

import asyncio
import threading
import time
from os import _exit, getenv
from socket import socket
from typing import Optional

import click
import hupper
import uvloop
from dotenv import load_dotenv

from . import __version__
from .environment import running_in_pytest
from .logs import configure_logging
from .metrics import no_metrics

# load .env environment variables as early as possible
if not running_in_pytest():  # pragma: no cover
    load_dotenv()

uvloop.install()


@click.command()
@click.option(
    "-l",
    "--log-level",
    type=click.Choice(["CRITICAL", "ERROR", "WARNING", "INFO", "DEBUG"]),
    default=None,
    help="INFO is not set, can also be set by the LOG_LEVEL environment variable.",
)
@click.option(
    "-d",
    "--dev",
    default=False,
    is_flag=True,
    help="Development mode, automatically reload bot when source changes",
)
@click.option(
    "-g",
    "--debug",
    default=False,
    is_flag=True,
    help="Enable detailed asyncio debugging",
)
@click.option(
    "-m",
    "--mock-games",
    default=False,
    is_flag=True,
    help="Produce mock game urls instead of real ones",
)
@click.option(
    "-a",
    "--api",
    default=False,
    is_flag=True,
    help="Start the API web server instead of the bot",
)
@click.option(
    "-p",
    "--port",
    type=click.INT,
    required=False,
    help="Use the given port number to serve the API",
)
@click.version_option(version=__version__)
def main(
    log_level: Optional[str],
    dev: bool,
    debug: bool,
    mock_games: bool,
    api: bool,
    port: Optional[int] = None,
) -> None:
    if dev:
        hupper.start_reloader("spellbot.main")

    from .settings import Settings

    # Ensure that configure_logging() is called as early as possible
    settings = Settings()
    level = log_level if log_level is not None else (getenv("LOG_LEVEL") or "INFO")
    configure_logging(level)

    import logging

    # ddtrace logging is awful and spammy
    ddtrace_logger = logging.getLogger("ddtrace")
    ddtrace_logger.propagate = False
    ddtrace_logger.setLevel(logging.CRITICAL)

    # When metrics are enabled, let's ensure that datadog-agent is running first...
    if not no_metrics():  # pragma: no cover
        import logging

        logger = logging.root
        conn: Optional[socket] = None
        connected = False

        logger.info("metrics enabled, checking for connection to statsd server...")
        while not connected:
            try:
                conn = socket()
                conn.connect(("127.0.0.1", 8126))
                logger.info("statsd server connection established")
                connected = True
                logger.info("waiting for statsd server to finish initialization...")
                time.sleep(5)
            except Exception as e:
                logger.info(f"statsd connection error: {e}, retrying...")
                time.sleep(1)
            finally:
                assert conn is not None
                conn.close()

    loop = asyncio.new_event_loop()
    if debug:
        loop.set_debug(True)
    if api:
        from .web import launch_web_server

        launch_web_server(settings, loop, port or settings.PORT)
        loop.run_forever()
    else:
        from . import SpellBot
        from .client import build_bot

        assert settings.BOT_TOKEN is not None
        bot = build_bot(loop=loop, mock_games=mock_games)

        if not debug and not running_in_pytest:
            # Sometimes the bot gets into a weird state where it doesn't properly
            # start up entirely; this causes tasks (see tasks_cog.py) to never begin
            # since those only start after the ready signal has been processed.
            # As such, let's check every 30 minutes if the bot is ready, because if
            # it isn't we can just kill it and have supervisord restart it for us.
            def killer(bot: SpellBot):
                while True:
                    time.sleep(30 * 60)  # 30 minute wait
                    if not bot.is_ready():
                        print("exiting due to readiness check failure")  # noqa: T201
                        _exit(1)
                    else:
                        print("readiness check passed")  # noqa: T201

            x = threading.Thread(target=killer, args=(bot,))
            x.start()

        bot.run(settings.BOT_TOKEN)
