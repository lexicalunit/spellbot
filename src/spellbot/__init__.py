from __future__ import annotations

from ._version import __version__
from .cli import main
from .client import SpellBot

__all__ = [
    "__version__",
    "main",
    "SpellBot",
]
