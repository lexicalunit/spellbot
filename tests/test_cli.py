from unittest.mock import ANY

from spellbot.cli import main


class TestCLI:
    def test_help(self, cli, runner, snapshot):
        result = runner.invoke(main, ["--help"])
        assert result.exit_code == 0
        assert result.output == snapshot
        cli.bot.run.assert_not_called()

    def test_run_api(self, cli, runner):
        result = runner.invoke(main, ["--api"])
        assert result.exit_code == 0
        cli.launch_web_server.assert_called_once_with(
            cli.settings,
            cli.loop,
            cli.settings.PORT,
        )
        cli.loop.run_forever.assert_called_once_with()

    def test_run_api_with_port(self, cli, runner):
        result = runner.invoke(main, ["--api", "--port", "200"])
        assert result.exit_code == 0
        cli.launch_web_server.assert_called_once_with(cli.settings, cli.loop, 200)
        cli.loop.run_forever.assert_called_once_with()

    def test_run_bot(self, cli, runner):
        result = runner.invoke(main, [])
        assert result.exit_code == 0
        assert result.output == ""
        cli.hupper.start_reloader.assert_not_called()
        cli.configure_logging.assert_called_once_with("INFO")
        cli.build_bot.assert_called_once_with(
            clean_commands=False,
            force_sync_commands=False,
            loop=cli.loop,
            mock_games=False,
        )
        cli.bot.run.assert_called_once_with("facedeadbeef")

    def test_run_bot_with_log_level(self, cli, runner):
        runner.invoke(main, ["--log-level", "DEBUG"])
        cli.configure_logging.assert_called_once_with("DEBUG")

    def test_run_bot_without_bot_token(self, cli, runner):
        cli.settings.BOT_TOKEN = None
        result = runner.invoke(main, [])
        assert result.exit_code == 1
        assert result.output == ""
        cli.hupper.start_reloader.assert_not_called()
        cli.configure_logging.assert_called_once_with("INFO")
        cli.build_bot.assert_not_called()
        cli.bot.run.assert_not_called()

    def test_run_bot_with_clean_commands(self, cli, runner):
        runner.invoke(main, ["--clean-commands"])
        cli.build_bot.assert_called_once_with(
            clean_commands=True,
            force_sync_commands=ANY,
            loop=ANY,
            mock_games=ANY,
        )

    def test_run_bot_with_force_sync_commands(self, cli, runner):
        runner.invoke(main, ["--sync-commands"])
        cli.build_bot.assert_called_once_with(
            clean_commands=ANY,
            force_sync_commands=True,
            loop=ANY,
            mock_games=ANY,
        )

    def test_run_bot_with_mock_games(self, cli, runner):
        runner.invoke(main, ["--mock-games"])
        cli.build_bot.assert_called_once_with(
            clean_commands=ANY,
            force_sync_commands=ANY,
            loop=ANY,
            mock_games=True,
        )

    def test_run_bot_with_debug(self, cli, runner):
        runner.invoke(main, ["--debug"])
        cli.loop.set_debug.assert_called_once_with(True)

    def test_run_bot_with_dev(self, cli, runner):
        runner.invoke(main, ["--dev"])
        cli.hupper.start_reloader.assert_called_once_with("spellbot.main")
