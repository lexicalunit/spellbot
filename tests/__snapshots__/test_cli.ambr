# name: TestCLI.test_help
  '
  Usage: main [OPTIONS]
  
  Options:
    -l, --log-level [CRITICAL|ERROR|WARNING|INFO|DEBUG]
                                    INFO is not set, can also be set by the
                                    LOG_LEVEL environment variable.
    -d, --dev                       Development mode, automatically reload bot
                                    when source changes
    -g, --debug                     Enable detailed asyncio debugging
    -m, --mock-games                Produce mock game urls instead of real ones
    -s, --sync-commands             Force sync all slash commands at startup
    -c, --clean-commands            Cleanup all unused slash commands at startup
    -a, --api                       Start the API web server instead of the bot
    -p, --port INTEGER              Use the given port number to serve the API
    --version                       Show the version and exit.
    --help                          Show this message and exit.
  
  '
---
